src/Microsoft.AspNetCore.Blazor.Server/Circuits/CircuitSynchronizationContext.cs

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Blazor.Server.Circuits
+{
+    [DebuggerDisplay("{_state,nq}")]
+    internal class CircuitSynchronizationContext : SynchronizationContext
+    {
+        private static readonly ContextCallback ExecutionContextThunk = (object state) =>
+        {
+            var item = (WorkItem)state;
+            item.SynchronizationContext.ExecuteSynchronously(null, item.Callback, item.State);
+        };
+
+        private static readonly Action<Task, object> BackgroundWorkThunk = (Task task, object state) =>
+        {
+            var item = (WorkItem)state;
+            item.SynchronizationContext.ExecuteBackground(item);
+        };
+
+        private readonly State _state;
+
+        public event UnhandledExceptionEventHandler UnhandledException;
+
+        public CircuitSynchronizationContext()
+            : this(new State())
+        {
+        }
+
+        private CircuitSynchronizationContext(State state)
+        {
+            _state = state;
+        }
+
+        public Task Invoke(Action action)
+        {
+            var completion = new TaskCompletionSource<object>();
+            Post(_ =>
+            {
+                try
+                {
+                    action();
+                    completion.SetResult(null);
+                }
+                catch (Exception exception)
+                {
+                    completion.SetException(exception);
+                }
+            }, null);
+
+            return completion.Task;
+        }
+
+        public Task InvokeAsync(Func<Task> asyncAction)
+        {
+            var completion = new TaskCompletionSource<object>();
+            Post(async (_) =>
+            {
+                try
+                {
+                    await asyncAction();
+                    completion.SetResult(null);
+                }
+                catch (Exception exception)
+                {
+                    completion.SetException(exception);
+                }
+            }, null);
+
+            return completion.Task;
+        }
+
+        public Task<TResult> Invoke<TResult>(Func<TResult> function)
+        {
+            var completion = new TaskCompletionSource<TResult>();
```

Comment:
Note we still require a separate task/completion here due to the way we surface exceptions and return values.